x-pack/plugins/observability_solution/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/accordion_waterfall.tsx

```diff
@@ -78,111 +85,166 @@ const StyledAccordion = euiStyled(EuiAccordion).withConfig({
+function Waterfall(props: WaterfallProps) {
+  const listRef = useRef<List>(null);
+  const rowSizeMapRef = useRef(new Map<number, number>());
+  const { traceList } = useWaterfallContext();
 
-  let children = waterfall.childrenByParentId[item.id] || [];
+  const onRowLoad = (index: number, size: number) => {
+    rowSizeMapRef.current.set(index, size);
+  };
 
-  const criticalPath = showCriticalPath ? getCriticalPath(waterfall) : undefined;
+  const getRowSize = (index: number) => {
+    // adds 1px for the border top
+    return rowSizeMapRef.current.get(index) || ACCORDION_HEIGHT + 1;
+  };
 
-  const criticalPathSegmentsById = groupBy(criticalPath?.segments, (segment) => segment.item.id);
+  const onScroll = ({ scrollTop }: { scrollTop: number }) => {
+    listRef.current?.scrollTo(scrollTop);
+  };
 
-  let displayedColor = item.color;
+  return (
+    <WindowScroller onScroll={onScroll}>
+      {({ registerChild }) => (
+        <AutoSizer disableHeight>
+          {({ width }) => (
+            <div data-test-subj="waterfall" ref={registerChild}>
+              <List
+                ref={listRef}
+                style={{ height: '100%' }}
+                itemCount={traceList.length}
+                itemSize={getRowSize}
+                height={window.innerHeight}
+                width={width}
+                itemData={{ ...props, traceList, onLoad: onRowLoad }}
+              >
+                {VirtualRow}
+              </List>
+            </div>
+          )}
+        </AutoSizer>
+      )}
+    </WindowScroller>
+  );
+}
 
-  if (showCriticalPath) {
-    children = children.filter((child) => criticalPathSegmentsById[child.id]?.length);
-    displayedColor = transparentize(0.5, item.color);
-  }
+const VirtualRow = React.memo(
+  ({
+    index,
+    style,
+    data,
+  }: ListChildComponentProps<
+    Omit<WaterfallNodeProps, 'node'> & {
+      traceList: IWaterfallNodeFlatten[];
+      onLoad: (index: number, size: number) => void;
+    }
+  >) => {
+    const { onLoad, traceList, ...props } = data;
 
-  const errorCount = waterfall.getErrorCount(item.id);
+    const ref = React.useRef<HTMLDivElement | null>(null);
+    useEffect(() => {
+      onLoad(index, ref.current?.getBoundingClientRect().height ?? ACCORDION_HEIGHT);
+    }, [index, onLoad]);
 
-  // To indent the items creating the parent/child tree
-  const marginLeftLevel = 8 * level;
+    return (
+      <div style={style} ref={ref}>
+        <WaterfallNode {...props} node={traceList[index]} />
+      </div>
+    );
+  },
+  areEqual
+);
 
-  function toggleAccordion() {
-    setIsOpen((isCurrentOpen) => !isCurrentOpen);
-  }
+const WaterfallNode = React.memo((props: WaterfallNodeProps) => {
+  const theme = useTheme();
+  const { duration, waterfallItemId, onClickWaterfallItem, timelineMargins, node } = props;
+  const { criticalPathSegmentsById, getErrorCount, updateTreeNode, showCriticalPath } =
+    useWaterfallContext();
```

Comment:
why do we use context?

Reply:
I didn't want to pass `criticalPath` and `criticalPathSegmentsById` via props, and also facilitate the `updateTreeNode` call, without having to pass it via props.

Reply:
Ok, I personally avoid context because it's easy to forget about wrapping things in context and there is no way to type check this, but no big deal.